sklearn/utils/estimator_checks.py

```diff
@@ -368,55 +375,201 @@ def check_transformer_pickle(name, Transformer):
+    """General tests for partial_fit method""" 
+
+    if not hasattr(Estimator, 'partial_fit'):
+        return
+
+    X, y = make_blobs(n_samples=50, random_state=1)
+    X -= X.min()
+
+    with warnings.catch_warnings(record=True):
+        estimator = Estimator()
+    set_fast_parameters(estimator)
+
+    # Check if returns self
+    if isinstance(estimator, ClassifierMixin):
+        classes = np.unique(y)
+        ret_val = estimator.partial_fit(X, y, classes=classes)
+
+        assert_equal(ret_val, estimator)
+
+    elif isinstance(estimator, RegressorMixin):
+        ret_val = estimator.partial_fit(X, y)
+        assert_equal(ret_val, estimator)
+  
     else:
-        alg.partial_fit(X, y)
+        ret_val = estimator.partial_fit(X)
+        assert_equal(ret_val, estimator)
+
+   
+def check_estimators_pfit_reset_when_fit(name, Estimator):
+    """Check if the estimators get reset when fit is applied after a 
+    partial fit.
+    """
+   
+    # Estimators are assumed to be reset when the transorm / predic result
+    # is similar to 1st fit call
+    # FIXME Remove MiniBatchKMeans from skiplist once it is fixed.
+    skip_list = ('MiniBatchKMeans')
+    if not hasattr(Estimator, 'partial_fit') or name in skip_list:
+        return
 
-    assert_raises(ValueError, alg.partial_fit, X[:, :-1], y)
+    X, y = make_blobs(n_samples=50, random_state=1)
+    X -= X.min()
 
+    with warnings.catch_warnings(record=True):
+        estimator = Estimator()
+    set_fast_parameters(estimator)
+    
+    classes = np.unique(y)
 
-def check_clustering(name, Alg):
+    # Initial call to fit
+    try:
+        estimator.fit(X)
+    except TypeError:
+        estimator.fit(X, y)
+
+    if hasattr(estimator, 'transform'):
+        result_first_fit = estimator.transform(X)
+    elif hasattr(estimator, 'predict'):
+        result_first_fit = estimator.predict(X)
+
+    # Run 2 iterations of full set of partial_fit with batches of size 10
+    for i in range(2):
+        X_s, y_s = shuffle(X, y, random_state=0)
+        for j in range(0, 50, 10):
+            if isinstance(estimator, ClassifierMixin):
+                estimator.partial_fit(X_s[j:10+j], y_s[j:10+j],
+                                      classes=classes)
+            elif isinstance(estimator, RegressorMixin):
+                estimator.partial_fit(X_s[j:10+j], y_s[j:10+j])
+            else:
+                estimator.partial_fit(X_s[j:10+j])
+         
+    emsg = "Change in features during 2nd fit raised an error.\n" 
+    try:
+        # if signature of fit is (X, y) raises TypeError
+        estimator.fit(X)
+        try:
+            # Test that change in features does not raise any error
+            estimator.fit(X[:, :-1])
+        except:
+            assert False, emsg + str(e)
+        # Finally call fit to reset the estimator
+        estimator.fit(X)
+    except TypeError:
+        try:
+            estimator.fit(X[:, :-1], y)
```

Comment:
Sorry, I didn't have time to read it closely. You're right, that's inapplicable. It just seemed strange that try-except should be necessary to catch an exception that we want raised. I guess the intention is that a particular message is shown. Perhaps that's unnecessary, as it makes the test much harder to read, but I'm not especially concerned as long as the code is clear (i.e. add a comment, please).

Also, never ever use `except:` without an exception type. Prefer `except Exception:` which will not catch `KeyboardInterrupt`.

>  there seems to be no check for the expected_exception...

And I agree, that looks like a bug in the `assert_raises_regex` implementation here, but shouldn't be true for the imported one I hope!!